dm-agent/src/main.rs

```diff
@@ -0,0 +1,190 @@
+    let config = mqtt_client::Config::new(DEFAULT_HOST, DEFAULT_PORT);
+    let mqtt = Arc::new(Client::connect(CLIENT_ID, &config).await?);
+    let filter = TopicFilter::new(SOURCE_TOPIC)?.qos(QoS::AtMostOnce);
+
+    let mut errors = mqtt.subscribe_errors();
+    let mut messages: MessageStream = mqtt.subscribe(filter).await?;
+
+    let mut interval =
+        tokio::time::interval(Duration::from_millis(DEFAULT_STATS_COLLECTION_INTERVAL));
+
+    let mut message_batch = MeasurementGrouper::new();
+
+    loop {
+        select! {
+            error = errors.next().fuse() => {
+                if let Some(err) = error {
+                    error!("{}", err);
+                }
+            }
+
+            maybe_message = messages.next().fuse() => {
+                if let Some(message) = maybe_message {
+                    if message_batch.is_empty() {
+                        let _ = message_batch.timestamp(&current_timestamp()).map_err(|err| error!("{}", err));
+                    }
+                    let _ = tedge_json_from_collectd_message(&mut message_batch, message).map_err(|err| error!("{}", err));
+                }
+            }
+
+            _result = interval.tick() => {
+                if !message_batch.is_empty() {
+                    handle_message_batch(mqtt.clone(), message_batch);
+                    message_batch = MeasurementGrouper::new();
+                }
+            }
+        }
+    }
+}
+
+fn tedge_json_from_collectd_message(
+    message_batch: &mut MeasurementGrouper,
+    message: Message,
+) -> anyhow::Result<()> {
+    let topic = message.topic.name.clone();
+    let split: Vec<&str> = topic.split('/').collect();
+    let group_measurement_key = split[split.len() - 2];
+    let measurement_key = split[split.len() - 1];
+
+    let payload = message.payload;
+    let payload = from_utf8(&payload)?;
+    let split: Vec<&str> = payload.split(':').collect();
+
+    let measurement_value = split[1].trim_end_matches(char::from(0)).parse::<f64>()?;
+    message_batch.measurement(
+        Some(group_measurement_key),
+        measurement_key,
+        measurement_value,
+    )?;
+
+    Ok(())
+}
+
+#[instrument(name = "message_batch_handler")]
+fn handle_message_batch(mqtt: Arc<Client>, message_batch: MeasurementGrouper) {
+    tokio::task::spawn(async move {
+        if let Err(err) = message_batch_processor(mqtt, message_batch).await {
+            error!("{}", err);
+        }
+    });
+}
+
+async fn message_batch_processor(
+    mqtt: Arc<Client>,
+    message_batch: MeasurementGrouper,
+) -> anyhow::Result<()> {
+    let mut tedge_json_serializer = ThinEdgeJsonSerializer::new();
+    message_batch.accept(&mut tedge_json_serializer)?;
+
+    let topic = Topic::new(TARGET_TOPIC).unwrap();
+    let tedge_message = Message::new(&topic, tedge_json_serializer.bytes()?);
+
+    mqtt.as_ref().publish(tedge_message).await?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
```

Comment:
There is no negative test case, there should be some added (wrong topic, wrong payload, etc.)